{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-03-09 14:21:45--  http://spades.bioinf.spbau.ru/students/2b/C24-698.fastq.gz\n",
      "Resolving spades.bioinf.spbau.ru (spades.bioinf.spbau.ru)... 194.85.238.21\n",
      "Connecting to spades.bioinf.spbau.ru (spades.bioinf.spbau.ru)|194.85.238.21|:80... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 97418857 (93M) [application/x-gzip]\n",
      "Saving to: <</Users/dashabalashova/ngs_institute/data/C24-698.fastq.gz>>\n",
      "\n",
      "/Users/dashabalasho 100%[===================>]  92.91M  1.36MB/s    in 73s     \n",
      "\n",
      "2019-03-09 14:22:58 (1.27 MB/s) - <</Users/dashabalashova/ngs_institute/data/C24-698.fastq.gz>> saved [97418857/97418857]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#!wget http://spades.bioinf.spbau.ru/students/2b/B22-730.fastq.gz -O ~/ngs_institute/data/B22-730.fastq.gz\n",
    "#!gunzip ~/ngs_institute/data/B22-730.fastq.gz\n",
    "\n",
    "\n",
    "\n",
    "#!wget http://spades.bioinf.spbau.ru/students/2b/DH10B-K12.fasta -O ~/ngs_institute/data/DH10B-K12.fasta\n",
    "\n",
    "\n",
    "!wget http://spades.bioinf.spbau.ru/students/2b/C24-698.fastq.gz -O ~/ngs_institute/data/C24-698.fastq.gz\n",
    "!gunzip ~/ngs_institute/data/C24-698.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bwa_index] Pack FASTA... 0.06 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[BWTIncCreate] textLength=9372274, availableWord=11906236\n",
      "[bwt_gen] Finished constructing BWT in 5 iterations.\n",
      "[bwa_index] 2.44 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.06 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.03 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.71 sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa index -a bwtsw /Users/dashabalashova/ngs_institute/data/DH10B-K12.fasta\n",
      "[main] Real time: 3.439 sec; CPU: 3.330 sec\n"
     ]
    }
   ],
   "source": [
    "#!bowtie2-build ~/ngs_institute/data/DH10B-K12.fasta DH10B-K12_index\n",
    "\n",
    "\n",
    "!bwa index -a bwtsw ~/ngs_institute/data/DH10B-K12.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[M::process] read 40748 sequences (10000257 bp)...\n",
      "[M::process] read 40830 sequences (10000256 bp)...\n",
      "[M::mem_process_seqs] Processed 40748 reads in 4.890 CPU sec, 4.895 real sec\n",
      "[M::process] read 40678 sequences (10000475 bp)...\n",
      "[M::mem_process_seqs] Processed 40830 reads in 4.816 CPU sec, 4.712 real sec\n",
      "[M::process] read 40934 sequences (10000123 bp)...\n",
      "[M::mem_process_seqs] Processed 40678 reads in 5.827 CPU sec, 5.759 real sec\n",
      "[M::process] read 40668 sequences (10000353 bp)...\n",
      "[M::mem_process_seqs] Processed 40934 reads in 4.382 CPU sec, 4.289 real sec\n",
      "[M::process] read 40714 sequences (10000220 bp)...\n",
      "[M::mem_process_seqs] Processed 40668 reads in 4.679 CPU sec, 4.746 real sec\n",
      "[M::process] read 40714 sequences (10000207 bp)...\n",
      "[M::mem_process_seqs] Processed 40714 reads in 5.148 CPU sec, 5.159 real sec\n",
      "[M::process] read 40584 sequences (10000468 bp)...\n",
      "[M::mem_process_seqs] Processed 40714 reads in 4.562 CPU sec, 4.606 real sec\n",
      "[M::process] read 40728 sequences (10000278 bp)...\n",
      "[M::mem_process_seqs] Processed 40584 reads in 4.869 CPU sec, 4.853 real sec\n",
      "[M::process] read 40746 sequences (10000126 bp)...\n",
      "[M::mem_process_seqs] Processed 40728 reads in 4.383 CPU sec, 4.319 real sec\n",
      "[M::process] read 40890 sequences (10000295 bp)...\n",
      "[M::mem_process_seqs] Processed 40746 reads in 4.851 CPU sec, 5.255 real sec\n",
      "[M::process] read 40758 sequences (10000106 bp)...\n",
      "[M::mem_process_seqs] Processed 40890 reads in 5.396 CPU sec, 6.349 real sec\n",
      "[M::process] read 40668 sequences (10000113 bp)...\n",
      "[M::mem_process_seqs] Processed 40758 reads in 5.342 CPU sec, 5.633 real sec\n",
      "[M::process] read 40834 sequences (10000508 bp)...\n",
      "[M::mem_process_seqs] Processed 40668 reads in 4.673 CPU sec, 4.610 real sec\n",
      "[M::process] read 40740 sequences (10000085 bp)...\n",
      "[M::mem_process_seqs] Processed 40834 reads in 4.827 CPU sec, 4.747 real sec\n",
      "[M::process] read 40748 sequences (10000509 bp)...\n",
      "[M::mem_process_seqs] Processed 40740 reads in 4.719 CPU sec, 4.637 real sec\n",
      "[M::process] read 33424 sequences (7905189 bp)...\n",
      "[M::mem_process_seqs] Processed 40748 reads in 4.684 CPU sec, 4.623 real sec\n",
      "[M::mem_process_seqs] Processed 33424 reads in 3.394 CPU sec, 3.331 real sec\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem /Users/dashabalashova/ngs_institute/data/DH10B-K12.fasta /Users/dashabalashova/ngs_institute/data/C24-698.fastq\n",
      "[main] Real time: 82.821 sec; CPU: 81.624 sec\n"
     ]
    }
   ],
   "source": [
    "#! cd ~/ngs_institute/data\n",
    "#!bowtie2 -x DH10B-K12_index -U ~/ngs_institute/data/B22-730.fastq -S B22-730.sam\n",
    "\n",
    "#!bowtie2 -x DH10B-K12_index -U ~/ngs_institute/data/C24-698.fastq -S C24-698.sam\n",
    "\n",
    "!bwa mem ~/ngs_institute/data/DH10B-K12.fasta ~/ngs_institute/data/C24-698.fastq > C24-698_bwamem.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "685406 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "0 + 0 secondary\n",
      "0 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "680266 + 0 mapped (99.25% : N/A)\n",
      "0 + 0 paired in sequencing\n",
      "0 + 0 read1\n",
      "0 + 0 read2\n",
      "0 + 0 properly paired (N/A : N/A)\n",
      "0 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (N/A : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "bamFilesList: ['C24-698.sorted.bam']\n",
      "binLength: 1000\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 1\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 1000\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n",
      "bamFilesList: ['C24-698.sorted.bam']\n",
      "binLength: 1\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 1\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 1\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n"
     ]
    }
   ],
   "source": [
    "#!samtools flagstat B22-730.sam\n",
    "#!samtools view -S -b B22-730.sam > B22-730.bam\n",
    "#!samtools sort B22-730.bam > B22-730.sorted.bam\n",
    "#!samtools index B22-730.sorted.bam\n",
    "#!bamCoverage -b B22-730.sorted.bam -bs 1000 -o B22-730_1000.bw --outFileFormat bedgraph\n",
    "#!bamCoverage -b B22-730.sorted.bam -bs 1 -o B22-730.bw --outFileFormat bedgraph\n",
    "\n",
    "!samtools flagstat C24-698.sam\n",
    "!samtools view -S -b C24-698.sam > C24-698.bam\n",
    "!samtools sort C24-698.bam > C24-698.sorted.bam\n",
    "!samtools index C24-698.sorted.bam\n",
    "!bamCoverage -b C24-698.sorted.bam -bs 1000 -o C24-698_1000.bw --outFileFormat bedgraph\n",
    "!bamCoverage -b C24-698.sorted.bam -bs 1 -o C24-698.bw --outFileFormat bedgraph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "686667 + 0 in total (QC-passed reads + QC-failed reads)\n",
      "0 + 0 secondary\n",
      "1261 + 0 supplementary\n",
      "0 + 0 duplicates\n",
      "682638 + 0 mapped (99.41% : N/A)\n",
      "0 + 0 paired in sequencing\n",
      "0 + 0 read1\n",
      "0 + 0 read2\n",
      "0 + 0 properly paired (N/A : N/A)\n",
      "0 + 0 with itself and mate mapped\n",
      "0 + 0 singletons (N/A : N/A)\n",
      "0 + 0 with mate mapped to a different chr\n",
      "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
      "bamFilesList: ['C24-698_bwamem.sorted.bam']\n",
      "binLength: 1000\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 1\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 1000\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n",
      "bamFilesList: ['C24-698_bwamem.sorted.bam']\n",
      "binLength: 1\n",
      "numberOfSamples: None\n",
      "blackListFileName: None\n",
      "skipZeroOverZero: False\n",
      "defaultFragmentLength: read length\n",
      "numberOfProcessors: 1\n",
      "verbose: False\n",
      "region: None\n",
      "bedFile: None\n",
      "minMappingQuality: None\n",
      "ignoreDuplicates: False\n",
      "chrsToSkip: []\n",
      "stepSize: 1\n",
      "center_read: False\n",
      "samFlag_include: None\n",
      "samFlag_exclude: None\n",
      "minFragmentLength: 0\n",
      "maxFragmentLength: 0\n",
      "zerosToNans: False\n",
      "smoothLength: None\n",
      "save_data: False\n",
      "out_file_for_raw_data: None\n",
      "maxPairedFragmentLength: 1000\n"
     ]
    }
   ],
   "source": [
    "!samtools flagstat C24-698_bwamem.sam\n",
    "!samtools view -S -b C24-698_bwamem.sam > C24-698_bwamem.bam\n",
    "!samtools sort C24-698_bwamem.bam > C24-698_bwamem.sorted.bam\n",
    "!samtools index C24-698_bwamem.sorted.bam\n",
    "!bamCoverage -b C24-698_bwamem.sorted.bam -bs 1000 -o C24-698_1000_bwamem.bw --outFileFormat bedgraph\n",
    "!bamCoverage -b C24-698_bwamem.sorted.bam -bs 1 -o C24-698_bwamem.bw --outFileFormat bedgraph"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
